greenplumpython/table.py

```diff
@@ -0,0 +1,140 @@
+import random
+import string
+from typing import Iterable, NamedTuple
+import sqlalchemy
+
+
+class Database:
+    def __init__(self, host: str = "localhost", db_name: str = "") -> None:
+        url = f"postgresql+psycopg2://{host}/{db_name}"
+        engine = sqlalchemy.create_engine(url)
+        self.conn = engine.connect()
+        self.name = db_name
+
+    def execute(self, query: str) -> Iterable:
+        return self.conn.execute(query)
+
+
+def database(host: str = "localhost", db_name: str = "") -> Database:
+    return Database(host=host, db_name=db_name)
+
+
+class Node:
+    def __init__(self) -> None:
+        pass
+
+
+class Table(Node):
+    def __init__(
+        self,
+        query: str,
+        parents: Iterable["Table"] = [],
+        table_name: str = None,
+        db: Database = None,
+    ) -> None:
+        self.query = query
+        self.parents = parents
+        if table_name is not None:
+            self.name = table_name
+        else:
+            self.name = "tb_" + "".join(
+                random.choice(string.ascii_lowercase) for _ in range(60)
+            )
+        if any(parents):
+            self.db = next(iter(parents)).db
+        else:
+            self.db = db
+
+    def __getitem__(self, key) -> "Table":
+        """
+        Returns
+        - a Column of the current Table if key is string, or
+        - a new Table from the current Table per the type of key:
+            - if key is a list, then SELECT a subset of columns, a.k.a. targets;
+            - if key is an Expr, then SELECT a subset of rows per the value of the Expr;
+            - if key is a slice, then SELECT a portion of consecutive rows
+        """
+        if isinstance(key, str):
+            return Column(key, self)
+        if isinstance(key, list):
+            return Table(
+                f"SELECT {','.join([str(target) for target in key])} FROM {self.name}",
+                parents=[self],
+            )
+        if isinstance(key, Expr):
+            return Table(f"SELECT * FROM {self.name} WHERE {str(key)}", parents=[self])
+        if isinstance(key, slice):
+            if key.step is not None:
+                raise NotImplementedError()
+            offset_clause = "" if key.start is None else f"OFFSET {key.start}"
+            return Table(
+                f"SELECT * FROM {self.name} LIMIT {key.stop} {offset_clause}",
+                parents=[self],
+            )
+
+    def list_lineage(self) -> Iterable["Table"]:
+        lineage = [self]
+        current = 0
+        while current < len(lineage):
+            for table in lineage[current].parents:
+                lineage.append(table)
```

Comment:
is logic here len(lineage) always==1?

Reply:
No. More items will be appended to the list as the loop goes.

Reply:
cool missed thanks